parser.cpp

```diff
@@ -67,7 +70,10 @@ Parser::Parser(
     if (p != std::string::npos)
         basename_ = std::string(
             basename_.begin(), basename_.begin() + static_cast<ptrdiff_t>(p));
-    lex_ = new Lexer(f);
+
+    Preprocessor cpp(defines_);
```

Comment:
If defines_ is empty, can we not invoke the preprocessor. That way, there won't be a hard requirement that a c preprocessor is required if the edl does not use #defines

Reply:
But often preprocessor directives are "Do something if this thing is _not_ defined. i.e. #ifndef

So I don't think this works.

Reply:
> But often preprocessor directives are "Do something if this thing is _not_ defined". i.e. #ifndef
> 
> So I don't think this works.

That's a great point.
In that case, we could just let the lexer recognize '#' as a token and do a first pass lexing to determine if an edl file requires preprocessor or not.

The reason I would like to be really sure before we commit to a hard dependency is that:
- We have ran into issues with preprocessor not being in path
https://github.com/openenclave/openenclave/blob/24a9b211e4ad7d4be46bf43297699459998a674f/tools/oeedger8r/intel/Preprocessor.ml#L55-L57
- sgx edger8r does not always error out if preprocessing fails:
https://github.com/intel/linux-sgx/blob/fdc9b33ee173a5e7743f20a2a9cb1dd8db114dca/sdk/edger8r/linux/Preprocessor.ml#L56

Reply:
Good point. I'll look into making those changes.

One question:
>  we could just let the lexer recognize '#' as a token and do a first pass lexing to determine if an edl file requires preprocessor or not.

My current implementation preprocesses the file prior to passing to the lexer. In this approach, the lexer would have to encounter a `#` character, preprocess the file, and then start over. Do you think that's ok?

I think about any alternative approaches.

Reply:
Oh actually an alternative to what I've said above is:
* Try to preprocess, if it doesn't work emit a warning and move on with the regular file
* If the lexer encounters a `#` character, error out with a message saying the file has not been preprocessed.

Reply:
> In this approach, the lexer would have to encounter a `#` character, preprocess the file, and then start over. Do you think that's ok?
>

That is totally fine. Even if we were to lex the same file multiple times, the performance impact would be unnoticable.

Reply:
So all I changed was to emit a warning when preprocessing fails and then just use the original file. So then if the preprocessor fails and the file didn't need it anyway then everything will work fine. And if it did need it then parsing will fail because `#` is invalid.